wgpu-hal/src/gles/wgl.rs

```diff
@@ -565,6 +582,43 @@ impl crate::Instance for Instance {
     }
 }
 
+impl super::Adapter {
+    /// Creates a new external adapter using the specified loader function.
+    ///
+    /// # Safety
+    ///
+    /// - The underlying OpenGL ES context must be current.
+    /// - The underlying OpenGL ES context must be current when interfacing with any objects returned by
+    ///   wgpu-hal from this adapter.
+    /// - The underlying OpenGL ES context must be current when dropping this adapter and when
+    ///   dropping any objects returned from this adapter.
```

Comment:
When designing `unsafe` APIs, there are two rules we should uphold:

1. Code inside `unsafe` has special considerations, and `unsafe` must be required _at any site where calling code can cause unsafety, if they do not follow the required safety con_.
2. Safe code (i.e., code _not_ in an `unsafe` block) should otherwise be able to rely on the Rust compiler to ensure safety; safe code should not need to be concerned with upholding safety invariants.

These are otherwise known as the Soundness Property:

> No matter what, Safe Rust can't cause Undefined Behavior.

See also the Rustonomicon's pages [Meet Safe and Unsafe](https://doc.rust-lang.org/nomicon/meet-safe-and-unsafe.html#meet-safe-and-unsafe) and [How Safe and Unsafe Interact](https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html#how-safe-and-unsafe-interact).

---

issue: The first condition seems to fit these criteria. However:

- The second condition here:

	> ```rust
	>     /// - The underlying OpenGL ES context must be current when interfacing with any objects returned by
	>     ///   wgpu-hal from this adapter.
	> ```

    …implies safety conditions to uphold in subsequent calls, rather than this one. An `unsafe` API design that adheres to rule (2), then, should make it `unsafe` to perform these subsequent calls, rather than impose this obligation here.
- The third condition:

	> ```
	>    /// - The underlying OpenGL ES context must be current when dropping this adapter and when
	>    ///   dropping any objects returned from this adapter.
	> ```

	…imposes a safety condition on `Drop`, which is a safe API. This violates rule (2); we need a different API for safely dropping this context, so users can't write safe Rust that might violate this condition.

Reply:
> and that's not something (IMO) we can ship

EGL shipped with exactly this though?

Reply:
@ErichDonGubler note that this is why I suggested to use `wglShareLists()` (and the EGL equivalent) before (like what GStreamer does) to make sure the contexts can be currented independently and on separate threads while sharing resources such as textures and surfaces for rendering.

Reply:
Ah, if that's the case, then let's file an issue and sort it out later. 👍🏻 I hadn't realized that we'd already shipped some APIs that don't respect the Soundness Property. I'll file a follow-up issue, unless there's one I'm not aware of you're already working against?

Reply:
Filed <https://github.com/gfx-rs/wgpu/issues/6211>.